enforcer/resources/owner_store.py

```diff
@@ -0,0 +1,126 @@
+import logging
+import threading
+import time
+from typing import Dict, Any, Optional, List
+
+from enforcer.env_vars import REPLICA_SET_CLEANUP_INTERVAL, REPLICA_SET_DELETION_WAIT
+from enforcer.model import PodOwner, RsOwner
+from enforcer.resources.kubernetes_resource_loader import KubernetesResourceLoader
+
+
+class OwnerStore:
+
+    def __init__(self):
+        self.rs_owners = self._load_rs_owners()
+        self._rs_owners_lock = threading.Lock()
+        self.cleanup_interval = REPLICA_SET_CLEANUP_INTERVAL
+        self._stop_event = threading.Event()
+        self._cleanup_thread = threading.Thread(target=self._periodic_cleanup, daemon=True)
+        self._cleanup_thread.start()
+
+    def _rs_key(self, rs_name: str, namespace: str) -> str:
+        return f"{namespace}/{rs_name}"
+
+    def _load_rs_owners(self) -> Dict[str, RsOwner]:
+        replica_sets_owners: List[RsOwner] = KubernetesResourceLoader.load_replicasets()
+        rs_owners: Dict[str, RsOwner] = {}
+        for owner in replica_sets_owners:
+            rs_owners[self._rs_key(owner.rs_name, owner.namespace)] = owner
+        return rs_owners
+
+    @staticmethod
+    def get_pod_name(metadata: Dict[str, Any]) -> str:
+        # if the pod's name is randomized, the name is under generateName
+        return metadata.get("name") or metadata.get("generateName")
+
+    def get_pod_owner(self, pod: Dict[str, Any]) -> Optional[PodOwner]:
+        metadata = pod.get("metadata", {})
+        owner_references = metadata.get("ownerReferences", [])
+        namespace: str = metadata.get("namespace")
+
+        try:
+            if not owner_references:  # pod has no owner, standalone pod. Return the pod
+                return PodOwner(
+                    kind="Pod", namespace=namespace, name=self.get_pod_name(pod)
+                )
+
+            # get only owners with controller == true
```

Comment:
why?

Reply:
only the controller is the one controlling the resource.
the others are used for garbage collection